src/GafferScene/MeshTangents.cpp

```diff
@@ -0,0 +1,155 @@
+	addChild( new StringPlug( "uvSet", Plug::In, "st" ) );
+	addChild( new StringPlug( "posPrimVarName", Plug::In, "P" ) );
+	addChild( new BoolPlug( "orthogonal", Plug::In, true ) );
+
+	// Fast pass-throughs for things we don't modify
+	outPlug()->attributesPlug()->setInput( inPlug()->attributesPlug() );
+	outPlug()->transformPlug()->setInput( inPlug()->transformPlug() );
+}
+
+MeshTangents::~MeshTangents()
+{
+}
+
+Gaffer::StringPlug *MeshTangents::uvSetPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+const Gaffer::StringPlug *MeshTangents::uvSetPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+Gaffer::StringPlug *MeshTangents::posPrimVarNamePlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+const Gaffer::StringPlug *MeshTangents::posPrimVarNamePlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+Gaffer::BoolPlug *MeshTangents::orthogonalPlug()
+{
+	return getChild<BoolPlug>( g_firstPlugIndex + 2 );
+}
+
+const Gaffer::BoolPlug *MeshTangents::orthogonalPlug() const
+{
+	return getChild<BoolPlug>( g_firstPlugIndex + 2 );
+}
+
+void MeshTangents::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	SceneElementProcessor::affects( input, outputs );
+
+	if( input == uvSetPlug() || input == posPrimVarNamePlug() || input == orthogonalPlug() )
+	{
+		outputs.push_back( outPlug()->objectPlug() );
+	}
+	else if( input == outPlug()->objectPlug() )
+	{
+		outputs.push_back( outPlug()->boundPlug() );
+	}
+}
+
+bool MeshTangents::processesBound() const
+{
+	return true;
+}
+
+void MeshTangents::hashProcessedBound( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	hashProcessedObject( path, context, h );
+}
+
+Imath::Box3f MeshTangents::computeProcessedBound( const ScenePath &path, const Gaffer::Context *context, const Imath::Box3f &inputBound ) const
+{
+	return inputBound;
+}
+
+bool MeshTangents::processesObject() const
+{
+	return true;
+}
+
+void MeshTangents::hashProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	uvSetPlug()->hash( h );
+	posPrimVarNamePlug()->hash( h );
+	orthogonalPlug()->hash( h );
+}
+
+IECore::ConstObjectPtr MeshTangents::computeProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::ConstObjectPtr inputObject ) const
+{
+	const MeshPrimitive *mesh = runTimeCast<const MeshPrimitive>( inputObject.get() );
+	if( !mesh )
+	{
+		return inputObject;
+	}
+
+	std::string uvSet = uvSetPlug()->getValue();
+	std::string posPrimVarName = posPrimVarNamePlug()->getValue();
+	bool ortho = orthogonalPlug()->getValue();
+
+	std::pair<PrimitiveVariable, PrimitiveVariable> tangentPrimvars = MeshAlgo::calculateTangents( mesh, uvSet, ortho, posPrimVarName);
+	MeshPrimitivePtr meshWithTangents = runTimeCast<MeshPrimitive>( mesh->copy() );
+
+	meshWithTangents->variables["Tangent"] = tangentPrimvars.first;
```

Comment:
I suppose I'm using the PR as mechanism to propose speculative code changes and follow the discussion, perhaps this is an abuse of the feature and frustrating.  

What does the phrase "along for the ride" mean in regards to primitive variables?

In retrospect what you describe sounds simple the for  the user to understand, general & cleaner API. I suppose I just went on *tangent* with regards to the derivative notation.

Should I wait for further discussion from other parties before implementing the following:

* add uTangentPlug() & vTangentPlug() which default to uTangent & vTangent. I assume were going to  be consistent using uv rather than st in cortex 10?